src/software/ai/hl/stp/stp.cpp

```diff
@@ -84,11 +85,16 @@ std::vector<std::unique_ptr<Intent>> STP::getIntentsFromCurrentPlay(const World&
     std::vector<std::unique_ptr<Intent>> intents;
     if (current_tactics)
     {
-        // Assign robots to tactics
-        auto assigned_tactics = assignRobotsToTactics(world, *current_tactics);
+        assignRobotsToTactics(world, *current_tactics);
 
-        for (const std::shared_ptr<Tactic>& tactic : assigned_tactics)
+        for (const std::shared_ptr<Tactic>& tactic : *current_tactics)
         {
+            // We only want to process tactics that have a robot assigned to them
+            if (!tactic->getAssignedRobot().has_value())
```

Comment:
We technically already have a case for this in the below if-else block. Because of how `tactics` return `nullptr` if they have no robot assigned, the logic below is ok, but I do not think it's handled in the best way. I'd recommend consolidating these checks at the top where you added the new check, and make sure to log there as well.

Reply:
Sorry, confused what you're asking for here. 
- I'm not comfortable relying on logic in the `Tactic` class here. 
- Which case in the below if-else do you think covers this?
- "I'd recommend consolidating these checks at the top ": what are "these checks"?

Reply:
On line, 98, the returned action will be a `nullptr` if the tactic does not have a robot assigned. This means we will not enter the following `if` cases for the action and intent, and will avoid performing any logic until we reach the `else` case on line 120. That is what I mean by that this is already covered. If tactics have no robot assigned, ultimately no logic will be run.

I also agree that not handling it explicitly at the start is not the correct way to do things. Now we are "handling" the case where the tactic has no robot both in your new check on line 93, and in the `else` case on line 120. These are the two checks I'm referring to that should be consolidated. We should only perform a single explicit check for robot assignment, at the beginning of the logic (where you currently added the check on line 93).

Is that more clear?

Reply:
Yup, removed the check on line 93 for now.